In [40]:
import requests
import re
import pandas as pd
import xlrd
from urllib.request import urlopen
from io import BytesIO
from zipfile import ZipFile
import xml.etree.ElementTree as ET
import time
import FinanceDataReader as fdr
from bs4 import BeautifulSoup

In [4]:
crtfc_key = open('crtfc_key.txt',mode='r').readline()
user_agent = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36'

# Corpcode

## Corpcode.xml 가져오기

In [13]:
url_corp = 'https://opendart.fss.or.kr/api/corpCode.xml?crtfc_key='+crtfc_key
# requests.get(url_corp).content.decode("UTF-8")
with urlopen(url_corp) as zipresp:
    with ZipFile(BytesIO(zipresp.read())) as zfile:
        zfile.extractall('Raw/Corpcode/corp_num')

## Corpcode.xml에서 기업코드 가져오기

In [3]:
tree = ET.parse('Raw/Corpcode/corp_num/CORPCODE.xml')
root = tree.getroot()

corp_df = pd.DataFrame(columns=['corp_code','corp_name','stock_code','modify_date'])
for company in root.iter('list'):
    stock_code = company.findtext('stock_code')
    stock_code = stock_code.strip()
    if stock_code:
        company_dict = {
            'corp_code':company.findtext('corp_code'),
            'corp_name':company.findtext('corp_name'),
            'stock_code':company.findtext('stock_code'),
            'modify_date':company.findtext('modify_date')
        }
        corp_df = corp_df.append(company_dict, ignore_index=True)
# corp_df

In [ ]:
# corp_df[corp_df['corp_name']=='BYC']
a = corp_df[corp_df['stock_code']=='168490'].iloc[0,0]
corp_df[corp_df['corp_code']==a]

# 필요 함수

## 함수생성을 위한 TEST

In [128]:
corp_code = stock_code_reset('005930')
bgn_de = "19990101"
url_api = "https://opendart.fss.or.kr/api/list.xml?crtfc_key=" + crtfc_key + "&corp_code=" + corp_code + "&bgn_de=" + bgn_de +"&last_reprt_at=Y" + "&pblntf_detail_ty=A001&pblntf_detail_ty=A002&pblntf_detail_ty=A003" + "&page_no=10&page_count=100"
dart_page = requests.get(url_api).content.decode("UTF-8")
rcp_no_list = re.findall(r'<rcept_no>(.*?)</rcept_no>', dart_page)
period_list = re.findall(r'<report_nm>(.*?)</report_nm>', dart_page)
year_list = list(map(year_reset, period_list))


In [146]:
period_list[80][-8:-1] +'_'+ period_list[35][-15:-10]

'2001.03_사업보고서'

In [208]:
rcp_no = '20110331002193'

In [55]:
url = 'http://dart.fss.or.kr/dsaf001/main.do?rcpNo={}'.format(rcp)
user_agent = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36'
resp = requests.get(url,headers={'user-agent':user_agent})
webpage = resp.content.decode("UTF-8")
webpage_clean = ['\n','\t',' ','(',')',"'",'"']
for clean in webpage_clean:
    webpage = webpage.replace(clean,'')
# try:
# webpage = re.findall(r"연결재무제표(.*?).xsd",webpage)[0]
# webpage = re.findall("부속명세서(.*?).xsd",webpage)[0]
# except:
# webpage = re.findall(r"재무제표(.*?).xsd",webpage)[0]
# webpage = re.findall(r"viewDoc(.*)",webpage)[0]
# dcm_eleid_no = webpage.split(',')[1:3]

In [212]:
webpage

'등,id:25,cls:text,listeners:{click:function{viewDoc20110331002193,2965731,25,1698770,139620,dart3'

In [58]:
eleid

9066

In [56]:
soup = BeautifulSoup(resp.content, 'html.parser')
soup


<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">

<html>
<head>
<title>
삼성전자/분기보고서/2005.11.14
</title>
<meta content="IE=EmulateIE8" http-equiv="X-UA-Compatible"/>
<meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>
<style type="text/css">
    .x-window-dlg .ext-mb-download {
        background:transparent url(images/download.gif) no-repeat top left;
        height:46px;
    }
</style>
<link charset="utf-8" href="/css/viewer.css" rel="stylesheet" type="text/css"/>
<script src="/js/prototype.js" type="text/javascript"></script>
<script src="/js/jquery/jquery-all.js" type="text/javascript"></script>
<!-- 2014.03.11 ext 3.4 -->
<!--[if lte IE 8]><link rel="stylesheet" type="text/css" href="/js/ext-viewer/resources/css/ext-all-ie8.css" ><![endif]-->
<!--[if (gte IE 9)|!(IE)]><!--><link href="/js/ext-viewer/resources/css/ext-all.css" rel="stylesheet" type="text/css"/><!--<![endif]-->
<script src="/js/ext-viewer/adapter/

In [90]:
# making_rcp_df(stock_code_reset('036670'))
making_rcp_df(stock_code_reset('005930'))

20210309000744
20201116001248
20200814001766
20200515001451
20200330003851
20191114001273
20190814002218
20190515001605
20190401004781
20181114001530
20180814001113
20180515001699
20180402005019
20171114001694
20170814001832
20170515003806
20170331004518
20161114001491
20160816001779
20160516003174
20160330003536
20151116000976
20150817000859
20150515001379
20150331002915
20141114000755
20140814000743
20140515001057
20140331002427
20131114000520
20130814000618
20130515000677
20130401003031
20121114000975
20120814001232
20120330002110
20120515001281
20111129000501
20110916000072
20110331002193
20110530000628
20101129000419
20100830000360
20100331001680
20100531001169
20091116000492
20090814001311
20090331002550
20090515001705
20081114000862
20080814001491
20080331002164
20080515001670
20071114001227
20070814001689
20070330001989
20070515001702
20061114001806
20060822000056
20060331001832
20060515002848
20051114001373
20050816000742
20050331002046
20050513002190
20041115001020
2004081300

stock_code description         2020.12         2020.09         2020.06  \
0     005930      rcp_no  20210309000744  20201116001248  20200814001766   
1     005930      dcm_no         7861003         7549051         7446167   
2     005930       eleid              14              13              13   

          2020.03         2019.12         2019.09         2019.06  \
0  20200515001451  20200330003851  20191114001273  20190814002218   
1         7341446         7202455         6958001         6846651   
2              13              13              13              13   

          2019.03  ...         2001.09         2001.06         2000.12  \
0  20190515001605  ...  20011114000621  20010814001128  20010331000297   
1         6738798  ...          263890          237424          178927   
2              13  ...            null            null            null   

          2001.03         2000.09         2000.06         1999.12  \
0  20010515000548  20001114000729  20000814000482  20000330000796   
1          206571          113424           89871           41243   
2            null            null            null            null   

          2000.03         1999.06         1998.12  
0  20000515000739  19990824000003  19990626000007  
1           60524           22086           15402  
2            null            null            null  

[3 rows x 89 columns]

## 실행 함수

In [5]:
def making_rcp_df(corp_code):
    bgn_de = "19990101"
    end_de = "20210410"
#     bgn_de = "20200101"
    url_api = "https://opendart.fss.or.kr/api/list.xml?crtfc_key=" + crtfc_key + "&corp_code=" + corp_code + "&bgn_de=" + bgn_de +"&end_de="+end_de+"&last_reprt_at=Y" + "&pblntf_detail_ty=A001&pblntf_detail_ty=A002&pblntf_detail_ty=A003" + "&page_no=10&page_count=100"
    dart_page = requests.get(url_api).content.decode("UTF-8")
    rcp_no_list = re.findall(r'<rcept_no>(.*?)</rcept_no>', dart_page)
    period_list = re.findall(r'<report_nm>(.*?)</report_nm>', dart_page)
    year_list = list(map(year_reset, period_list))
    
#     dcm_no_list = list(map(making_dcm, rcp_no_list))
    dcm_eleid_no = list(map(making_dcm_eleid,rcp_no_list))
    dcm_no_list = pd.DataFrame(dcm_eleid_no)[0].to_list()
    eleid_list = pd.DataFrame(dcm_eleid_no)[1].to_list()
    
    stock_code = corp_code_reset(corp_code)
    
    rcp_no_list.insert(0,stock_code)
    rcp_no_list.insert(1,'rcp_no')
    year_list.insert(0,'stock_code')
    year_list.insert(1,'description')
    dcm_no_list.insert(0,stock_code)
    dcm_no_list.insert(1,'dcm_no')
    eleid_list.insert(0,stock_code)
    eleid_list.insert(1,'eleid')
    
#     rcp_df = pd.DataFrame(data=[rcp_no_list,dcm_no_list],columns=year_list)
    rcp_df = pd.DataFrame(data=[rcp_no_list,dcm_no_list,eleid_list],columns=year_list)
    rcp_df = rcp_df.loc[:, ~rcp_df.columns.duplicated(
        keep='first')]  #한 해에 여러개의 보고서가 나오면 최근거만 가져옴
    return rcp_df

# def making_dcm (rcp_no):
#     url = 'http://dart.fss.or.kr/dsaf001/main.do?rcpNo={}'.format(rcp_no)
#     user_agent = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36'
#     resp = requests.get(url,headers={'user-agent':user_agent})
#     webpage = resp.content.decode("UTF-8")
#     dcm_no = re.findall(r"{}', '(.*?)',".format(rcp_no),webpage)[0]
#     time.sleep(0.5)
#     return dcm_no

def making_dcm_eleid (rcp_no):
#     print(rcp_no)
    url = 'http://dart.fss.or.kr/dsaf001/main.do?rcpNo={}'.format(rcp_no)
    user_agent = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36'
    resp = requests.get(url,headers={'user-agent':user_agent})
    webpage = resp.content.decode("UTF-8")
    webpage_clean = ['\n','\t',' ','(',')',"'",'"']
    for clean in webpage_clean:
        webpage = webpage.replace(clean,'')
    search_words = ['연결재무제표','재무제표','부속명세서']
    for search_word in search_words:
        try :
            webpage = re.findall(search_word+"(.*?).dart",webpage)[0]
            break
        except:
            continue
    webpage = re.findall(r"viewDoc(.*)",webpage)[0]
    dcm_eleid_no = webpage.split(',')[1:3]
    time.sleep(0.5)
    return dcm_eleid_no

def corp_code_reset(corp_code):  #에러가 발생할 경우 회사 코드 불러오기 실핸한다.
    stock_code = corp_df[corp_df['corp_code'] == corp_code].iloc[0, 2]
    return stock_code

def stock_code_reset(stock_code):
    corp_code = corp_df[corp_df['stock_code'] == stock_code].iloc[0, 0]
    return corp_code

def year_reset(title):
    try:
        year = period_list[35][-8:-1] +'_'+ period_list[35][-15:-10]
#         year = [i for i in re.findall('\(([^)]+)', title) if i][0]
    except:
        year = '0'
    return year


# rcp, dcm 가져오기

## stock_code 가져오기

In [92]:
# stock_df = fdr.StockListing('KOSPI')
stock_code_list = ['005930','005380','017670']

## rcp, dcm, eleid 만들기

In [152]:
# start = time.time()
rcp_df = pd.DataFrame()
indexerror_list=[]
valueerror_list=[]
for i in stock_code_list:
    try :
        rcp_df_temp = making_rcp_df(stock_code_reset(i))
        rcp_df = pd.concat([rcp_df,rcp_df_temp],axis=0)
    except IndexError:
        indexerror_list.append(i)
    except ValueError:
        valueerror_list.append(i)
    except Exception as e:
        print (i,e)
rcp_df.reset_index(drop=True,inplace=True)
# print(time.time()-start)
rcp_df.to_csv('Raw/rcp_dcm/rcp_df.csv', index=True, encoding='utf-8-sig')
# IndexError : single positional indexer is out-of-bounds : 우선주 등 재무제표 없는 경우
# list index out of range : 보고서 제목에 년도가 안써있는 경우 -> 0으로 쓰는것으로 수정
# ValueError : Length mismatch: Expected axis has 0 elements, new values have 1 elements : 리츠 등 재무제표 없는 경우

005930 HTTPSConnectionPool(host='opendart.fss.or.kr', port=443): Max retries exceeded with url: /api/list.xml?crtfc_key=f019243de247f120e3db654abec1d1ee6398eb72&corp_code=00126380&bgn_de=19990101&end_de=20210410&last_reprt_at=Y&pblntf_detail_ty=A001&pblntf_detail_ty=A002&pblntf_detail_ty=A003&page_no=10&page_count=100 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001FA6B3E5340>: Failed to establish a new connection: [WinError 10060] 연결된 구성원으로부터 응답이 없어 연결하지 못했거나, 호스트로부터 응답이 없어 연결이 끊어졌습니다'))


KeyboardInterrupt: 

In [119]:
rcp_df

stock_code description         2020.12         2020.09         2020.06  \
0     005930      rcp_no  20210309000744  20201116001248  20200814001766   
1     005930      dcm_no         7861003         7549051         7446167   
2     005930       eleid              14              13              13   
3     005380      rcp_no  20210316000986  20201116001588  20200814002203   
4     005380      dcm_no         7880318         7549889         7447766   
5     005380       eleid              14              13              13   
6     017670      rcp_no  20210317000804  20201116000869  20200814002234   
7     017670      dcm_no         7884796         7548156         7447875   
8     017670       eleid              14              13              13   

          2020.03         2019.12         2019.09         2019.06  \
0  20200515001451  20200330003851  20191114001273  20190814002218   
1         7341446         7202455         6958001         6846651   
2              13              13              13              13   
3  20200515002179  20200330004041  20191114001732  20190814002230   
4         7343256         7203672         6959184         6846694   
5              13              13              13              13   
6  20200515002457  20200330003443  20191114001848  20190814002530   
7         7344012         7199777         6959482         6847756   
8              13              13              13              13   

          2019.03  ...         2001.09         2001.06         2000.12  \
0  20190515001605  ...  20011114000621  20010814001128  20010331000297   
1         6738798  ...          263890          237424          178927   
2              13  ...            8000            7801           11143   
3  20190515002011  ...  20011113000302  20010816000050  20010330000582   
4         6739815  ...          262569          237756          174571   
5              13  ...            3357            3310            3237   
6  20190515001803  ...  20011114000904  20010814000621  20010331000096   
7         6739285  ...          264246          236424          178107   
8              13  ...            9754            9697            7477   

          2001.03         2000.09         2000.06         1999.12  \
0  20010515000548  20001114000729  20000814000482  20000330000796   
1          206571          113424           89871           41243   
2            7691           10786            7502            7334   
3  20010514000734  20001114000003  20000814000121  20000503000073   
4          205881          112238           89163           55439   
5            3214            3158            3222              89   
6  20010515000510  20001114000690  20000814000992  20000330000523   
7          206530          113330           90617           40681   
8            9438            5504            6297            3726   

          2000.03         1999.06         1998.12  
0  20000515000739  19990824000003  19990626000007  
1           60524           22086           15402  
2            9521            6700            8883  
3  20000515000428  19991209000001  19990407000005  
4           60064           24529           10115  
5            3117            2696            2490  
6  20000515000257  19990813000219  19990610000004  
7           59813           20448           13386  
8            4597            3806            3242  

[9 rows x 89 columns]

## rcp, dcm 불러오기

In [6]:
def six_degree (stock_code):
    stock_code_return = '%06d'%(stock_code)
    return stock_code_return

In [7]:
rcp_df = pd.read_csv('Raw/rcp_dcm/rcp_df.csv')
# rcp_df = rcp_df.set_index(rcp_df.columns[0])
# rcp_df.index.name = None
rcp_df = rcp_df.drop(columns=rcp_df.columns[0])
rcp_df.reset_index(drop=True,inplace=True)
# column 지정
# rcp_df.columns = [c.columns.map(lambda x : x[:2]), c.iloc[0], c.iloc[1]]
# c = c.iloc[2:]
# 종목명 6자리로 변경
# for i in range(len(rcp_df.index)):
#     rcp_df.rename(index={rcp_df.index[i]:'%06d'%(rcp_df.index[i])}, inplace = True)
rcp_df['stock_code']=list(map(six_degree, rcp_df['stock_code']))

rcp_df

stock_code description         2020.12         2020.09         2020.06  \
0     005930      rcp_no  20210309000744  20201116001248  20200814001766   
1     005930      dcm_no         7861003         7549051         7446167   
2     005930       eleid              14              13              13   
3     005380      rcp_no  20210316000986  20201116001588  20200814002203   
4     005380      dcm_no         7880318         7549889         7447766   
5     005380       eleid              14              13              13   
6     017670      rcp_no  20210317000804  20201116000869  20200814002234   
7     017670      dcm_no         7884796         7548156         7447875   
8     017670       eleid              14              13              13   

          2020.03         2019.12         2019.09         2019.06  \
0  20200515001451  20200330003851  20191114001273  20190814002218   
1         7341446         7202455         6958001         6846651   
2              13              13              13              13   
3  20200515002179  20200330004041  20191114001732  20190814002230   
4         7343256         7203672         6959184         6846694   
5              13              13              13              13   
6  20200515002457  20200330003443  20191114001848  20190814002530   
7         7344012         7199777         6959482         6847756   
8              13              13              13              13   

          2019.03  ...         2001.09         2001.06         2000.12  \
0  20190515001605  ...  20011114000621  20010814001128  20010331000297   
1         6738798  ...          263890          237424          178927   
2              13  ...            8000            7801           11143   
3  20190515002011  ...  20011113000302  20010816000050  20010330000582   
4         6739815  ...          262569          237756          174571   
5              13  ...            3357            3310            3237   
6  20190515001803  ...  20011114000904  20010814000621  20010331000096   
7         6739285  ...          264246          236424          178107   
8              13  ...            9754            9697            7477   

          2001.03         2000.09         2000.06         1999.12  \
0  20010515000548  20001114000729  20000814000482  20000330000796   
1          206571          113424           89871           41243   
2            7691           10786            7502            7334   
3  20010514000734  20001114000003  20000814000121  20000503000073   
4          205881          112238           89163           55439   
5            3214            3158            3222              89   
6  20010515000510  20001114000690  20000814000992  20000330000523   
7          206530          113330           90617           40681   
8            9438            5504            6297            3726   

          2000.03         1999.06         1998.12  
0  20000515000739  19990824000003  19990626000007  
1           60524           22086           15402  
2            9521            6700            8883  
3  20000515000428  19991209000001  19990407000005  
4           60064           24529           10115  
5            3117            2696            2490  
6  20000515000257  19990813000219  19990610000004  
7           59813           20448           13386  
8            4597            3806            3242  

[9 rows x 89 columns]

# 재무제표 다운로드

## Error_case
- 연결재무제표가 제출예정인 경우, rcp = 20070330001989
- 연결재무제표에 '재무상태표'등 항목명이 안써있는 경우 : 20051114001373 (


In [48]:
stock_code='005930'
eleid='2006.12'
rcp,dcm,eleid = rcp_df[rcp_df['stock_code']=='005930']['2005.09'].values

In [62]:
rcp

20051114001373

In [28]:
# error_df.to_csv('error_df.csv')
# error_df=pd.read_csv('error_df.csv',encoding='utf-8',index_col=0).astype(int)
error_df

0           1           2           3           4  \
stock_code        5930        5930        5930        5930        5930   
year              2006        2005        2005        2005        2004   
rcp        -2147483648 -2147483648 -2147483648 -2147483648 -2147483648   
dcm            1574002     1321907     1202395     1156947     1049370   
eleid             8684        9026        9066        9014        9221   

                     5           6           7           8           9  \
stock_code        5930        5930        5380        5380        5380   
year              2004        2004        2006        2005        2004   
rcp        -2147483648 -2147483648 -2147483648 -2147483648 -2147483648   
dcm             947352      908765     1574844     1313608     1046393   
eleid            15437       15374        8274        7925        7346   

                    10          11          12          13  
stock_code        5380        5380        5380        5380  
year              2002        2001        2000        1998  
rcp        -2147483648 -2147483648 -2147483648 -2147483648  
dcm             570011      343847      174571       10115  
eleid             8091        4228        3237        2490

In [255]:
error_df = pd.DataFrame(index = ['stock_code','year','rcp','dcm','eleid'])

for no_of_stock in range(int(len(rcp_df.index)/3)):
    rcp_list_temp = rcp_df.iloc[3*no_of_stock,2:].values
    dcm_list_temp = rcp_df.iloc[3*no_of_stock+1,2:].values
    eleid_list_temp = rcp_df.iloc[3*no_of_stock+2,2:].values
    year_list_temp = rcp_df.columns[2:]
    stock_code_temp = rcp_df.iloc[3*no_of_stock,0]
    for year, rcp, dcm, eleid in zip(year_list_temp,rcp_list_temp,dcm_list_temp,eleid_list_temp):
        try:
            download_ftable_html(stock_code_temp,year,rcp,dcm,eleid)
        except:
            no_error=len(error_df.columns)
            error_df[no_error] = [stock_code_temp,year,rcp,dcm,eleid]

In [78]:
def download_ftable_excel(stock_code,year,rcp,dcm):
    url_download = 'http://dart.fss.or.kr/pdf/download/excel.do?rcp_no={}&dcm_no={}&lang=ko'.format(rcp,dcm)
    user_agent = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36'
    resp_download = requests.get(url_download, headers = {'user-agent' : user_agent}, verify=False)
    table_download = BytesIO(resp_download.content)
    sheet_names = pd.ExcelFile(table_download).sheet_names
    sheets_group = [['연결 재무상태표','재무상태표','대차대조표'],['연결 현금흐름표','현금흐름표','현금흐름표'],['연결 포괄손익계산서','포괄손익계산서','손익계산서']]
    for sheets in sheets_group:
        i=0
        while True:
            if sheets[i] not in sheet_names:
                i=i+1
                if i > len(sheets):
                    print("재무상태표를 찾을 수 없습니다")
                    break
            else: 
                data_download = pd.read_excel(table_download, sheet_name = sheets[i])
                try:
                    data_download.to_csv("Raw/Dart_Crawling/"+stock_code+"_"+year+"_"+sheets[1]+".csv", encoding='euc-kr')
                except UnicodeEncodeError:
                    # 최근것은 euc-kr로 encoding되어 있으나, 2010년이전에는 utf-16으로 encoding된것으로 보임.
                    # 다만, utf-16에서 to_csv할 때 sep=','이 적용되지 않아 다운로드된 파일의 글씨는 깨지지만 utf-8로 사용.
                    data_download.to_csv("Raw/Dart_Crawling/"+stock_code+"_"+year+"_"+sheets[1]+".csv", encoding='utf-8')
                break

In [51]:
def download_ftable_html(stock_code,year,rcp,dcm,eleid):
    url = 'http://dart.fss.or.kr/report/viewer.do?rcpNo={}&dcmNo={}&eleId={}&offset=687340&length=115750&dtd=dart3.xsd'.format(int(rcp),int(dcm),int(eleid))
    table = pd.read_html(url)
    sheets_group = [['연결재무상태표','재무상태표','대차대조표'],['연결현금흐름표','현금흐름표','현금흐름표'],['연결포괄손익계산서','포괄손익계산서','손익계산서']]
    for sheets in sheets_group:
        for sheet in sheets:
            find=0
            for i in range(len(table)):
                if sheet in str(table[i].iloc[0,0]).replace(' ',''):
                    unit = table[i].iloc[-1,0]
                    table_temp = table[i+1]
                    table_temp['unit'] = unit
                    table_temp.to_csv("Raw/Dart_Crwaling_Table/"+str(stock_code)+"_"+str(year)+"_"+sheets[1]+".csv", encoding='utf-8-sig')
                    find=1
                    break
            if find == 1:
                break


In [21]:
stock_code, year,rcp, dcm, eleid = error_df.iloc[:,0].values
# error_df.iloc[:,0].values

In [52]:
download_ftable_html(stock_code, year,rcp, dcm, eleid)

IndexError: list index out of range

In [50]:
import webbrowser
url = 'http://dart.fss.or.kr/dsaf001/main.do?rcpNo={}'.format(rcp)
webbrowser.open(url)

True

In [59]:
url = 'http://dart.fss.or.kr/report/viewer.do?rcpNo={}&dcmNo={}&eleId={}&offset=687340&length=115750&dtd=dart3.xsd'.format(int(rcp),int(dcm),int(eleid))
table = pd.read_html(url)

In [61]:
# for i in range(len(table)):
#     print(table[i].iloc[0,0])
# str(table[1].iloc[0,0]).replace(' ','')
table[1]

구 분      제 36 기      제 35 기      제 34 기    제 33 기      제 32 기
0        [유동자산]    31991225    35555135    36164205  26870721    25796284
1         ㆍ당좌자산    26187579    30773930    31867298  23259989    21071408
2         ㆍ재고자산     5803646     4781205     4296907   3610732     4724876
3        [고정자산]    37013400    32486606    28790781  25244157    20625220
4         ㆍ투자자산    12506482    12550946    12164350   9551003     6072335
5         ㆍ유형자산    23962396    19470332    16277269  15401352    14299532
6         ㆍ무형자산      544522      465328      349162    291802      253353
7          자산총계    69004625    68041741    64954986  52114878    46421504
8        [유동부채]    26107583    28020512    28351235  23561619    24862214
9        [고정부채]     6496761     9861517    10282405   8154740     4773794
10         부채총계    32604344    37882029    38633640  31716359    29636008
11       외부주주지분     1956715      669122     1179781    908597      534930
12        [자본금]      897514      895241      889147    882117      881675
13    [연결자본잉여금]     6239586     6242269     5931788   5723591     5819652
14    [연결이익잉여금]    30576954    24415681    20322113  13341928    10719539
15     [연결자본조정]  △3,270,488  △2,062,601  △2,001,483  △457,714  △1,170,301
16         자본총계    36400281    30159712    26321346  20398519    16785495
17          매출액    81963009    64817456    58504801  46443768    43527806
18         영업이익    11760746     6296227     9129485   3951428     9060340
19         경상이익    12376158     6429560     9569039   3834429     8573099
20       총당기순이익    10168873     5353068     7324642   3370912     6192090
21      연결당기순이익    10789535     5962247     7052835   3055066     6002881
22  연결에 포함된 회사수        104개        101개        102개     101 개       100 개

# ETC

In [85]:
# i=4
for i in range(10):
    while True:
        if i != 6:
            if i == 8:
                print('break')
                break
            i=i+1
        else:
            print(i)
            break

6
6
6
6
6
6
6
break
break


KeyboardInterrupt: 